content/blog/2023/2023-09-16-create-openfeign-http-client.md

```diff
@@ -0,0 +1,503 @@
+      case 404:
+        return new RuntimeException(messageStr.isEmpty() ? "Not found" : messageStr);
+      default:
+        return defaultErrorDecoder.decode(methodKey, response);
+    }
+  }
+}
+```
+All responses with HTTP status other than `HTTP 2xx` range, for example `HTTP 400`, will trigger the ErrorDecoder's `decode` method. 
+In this overridden `decode` method, we can handle the response, wrap the failure into a custom exception or perform any additional processing.
+
+
+We can even retry the request again by throwing a `RetryableException`. 
+This will invoke the registered `Retryer`. [Retryer](#configuring-retryer) is explained in detail in the advanced techniques.
+
+You can see it in action by running `givenNegativeDivisorDivisionReturnsError` test in the example code shared on [Github](#example-code).
+
+## Advanced Techniques
+
+### Integrating Encoder/Decoder
+Encoder and decoder are **used to encode/decode the request and response data respectively**. We select these depending on the content type of the request and response. For example, [Gson](https://github.com/OpenFeign/feign/blob/master/gson) or [Jackson](https://github.com/OpenFeign/feign/blob/master/jackson)
+ can be used for JSON data.
+
+Here is an example showing how to use `Jackson` encoder and decoder.
+```java
+final CalculatorService target = Feign.builder()
+                                      .encoder(new JacksonEncoder())
+                                      .decoder(new JacksonDecoder())
+                                      .target(CalculatorService.class, HOST);
+
+```
+
+### Changing HTTP Client
+By default, it uses Feign HTTP client. The motivation behind changing the default HTTP client of Feign, from the original Apache HTTP Client to other libraries like OkHttp, is primarily driven by the need for **better performance, improved features, and enhanced compatibility with modern HTTP standards**.
+
+Now let us see how to override the HTTP client.
+```java
+final CalculatorService target = Feign.builder()
+                                      .client(new OkHttpClient())
+                                      .target(CalculatorService.class, HOST);
+
+```
+
+### Applying Circuit Breaker
+
+### Configuring Logger
+[SLF4JModule](https://github.com/OpenFeign/feign/blob/master/slf4j) is used to send Feign's logging to [SLF4J](http://www.slf4j.org/). With SLF4J, we can easily use a logging backend of our choice (Logback, Log4J, etc.)
+
+Here is an example about building the client:
+```java
+final CalculatorService target = Feign.builder()
+                                      .logger(new Slf4jLogger())
+                                      .logLevel(Level.FULL)
+                                      .target(CalculatorService.class, HOST);
+
+```
+To use SLF4J with Feign, add both the SLF4J module and an SLF4J binding of our choice to the classpath. 
+Then, configure Feign to use the Slf4jLogger:
+
+### Configuring Request Interceptors
```

Comment:
We don't want to "configure request interceptors", we want to "Intercept Requests"